dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/LoopAnalysis.java

```diff
@@ -0,0 +1,202 @@
+                continue;
+            }
+
+            final String loopName = labelInfo.loopName;
+            if (labelInfo.iterationIndex == 1) {
+                // We start a new loop
+                final LoopInfo loopInfo = new LoopInfo();
+                loopInfo.loopNumber = loopCounter++;
+                loopInfo.loopName = loopName;
+                loopInfo.thread = thread;
+                loopInfo.isUnrolled = true;
+                // NOTE: Inner loops may get copied while unrolling the outer loop.
+                // In that case, the inner loops all may have identical label names
+                // So this put-operation may overwrite an old entry, which is okay!
+                loopName2InfoMap.put(loopName, loopInfo);
+                loops.add(loopInfo);
+            }
+
+            final LoopInfo loopInfo = loopName2InfoMap.get(loopName);
+            if (labelInfo.isBound) {
+                final LoopIterationInfo lastIter = loopInfo.loopIterationInfos.get(loopInfo.loopIterationInfos.size() - 1);
+                // We assume the bound marker to be directly before the loop bounding event.
+                // NOTE: We consider neither the bound marker nor the bounding event to be part of this iteration.
+                lastIter.iterEnd = e.getPredecessor();
+            } else {
+                final LoopIterationInfo newIter = new LoopIterationInfo();
+                newIter.iterNumber = labelInfo.iterationIndex;
+                newIter.iterBegin = e;
+                newIter.containingLoop = loopInfo;
+                loopInfo.loopIterationInfos.add(newIter);
+                Verify.verify(newIter.iterNumber == loopInfo.loopIterationInfos.size());
+
+                if (loopInfo.loopIterationInfos.size() > 1) {
+                    final LoopIterationInfo prevIter = loopInfo.loopIterationInfos.get(loopInfo.loopIterationInfos.size() - 2);
+                    prevIter.iterEnd = e.getPredecessor(); // The previous iteration ends directly before the new starts
+                }
+            }
+        }
+
+        return ImmutableList.copyOf(loops);
+    }
+
+    // A loop-related label has the form
+    // "origLabel.loop/bound OR origLabel.loop/itr_N"
+    private LoopLabelInfo tryParseLoopLabel(Event eventToParse) {
+        if (!(eventToParse instanceof Label && ((Label)eventToParse).getName().contains(".loop"))) {
+            return null;
+        }
+
+        final Label label = (Label) eventToParse;
+        final String loopLabelSuffixSeparator = "/";
+        final int labelSuffixIndex = label.getName().lastIndexOf(loopLabelSuffixSeparator);
+        final String loopName = label.getName().substring(0, labelSuffixIndex);
+        final String labelSuffix = label.getName().substring(labelSuffixIndex + loopLabelSuffixSeparator.length());
+
+        final LoopLabelInfo info = new LoopLabelInfo();
+        info.loopName = loopName;
+        if (labelSuffix.contains("bound")) {
```

Comment:
Same discussion as above: I'm not a fan of having `Label` be filled with loop-specific code., in particular because the distinction between "loop labels" and "regular labels" is quite weak, given that after unrolling there are no more loops and all jumps/labels are structurally equivalent.